File: VizOs2/VizOs/frontend/js/roundrobin.js
// Round Robin CPU Scheduling Algorithm
class RoundRobin {
    static simulate(processes, timeQuantum) {
        // Sort processes by arrival time
        const sortedProcesses = [...processes].sort((a, b) => a.arrival - b.arrival);
        
        let currentTime = 0;
        const ganttChart = [];
        const processResults = [];
        const readyQueue = [];
        let processIndex = 0;
        
        // Initialize process data
        const processData = sortedProcesses.map(p => ({
            ...p,
            remainingTime: p.burst,
            startTime: -1,
            completionTime: -1,
            turnaroundTime: -1,
            waitingTime: -1,
            responseTime: -1
        }));
        
        while (processIndex < processData.length || readyQueue.length > 0) {
            // Add processes that have arrived to ready queue
            while (processIndex < processData.length && 
                   processData[processIndex].arrival <= currentTime) {
                readyQueue.push(processData[processIndex]);
                processIndex++;
            }
            
            if (readyQueue.length === 0) {
                // No processes ready, advance time to next arrival
                if (processIndex < processData.length) {
                    currentTime = processData[processIndex].arrival;
                    continue;
                } else {
                    break;
                }
            }
            
            // Get next process from ready queue
            const process = readyQueue.shift();
            
            // Set start time if this is the first time the process runs
            if (process.startTime === -1) {
                process.startTime = currentTime;
                process.responseTime = process.startTime - process.arrival;
            }
            
            // Execute for time quantum or remaining time, whichever is smaller
            const executionTime = Math.min(timeQuantum, process.remainingTime);
            const startExecution = currentTime;
            const endExecution = startExecution + executionTime;
            
            // Add to Gantt chart
            ganttChart.push({
                name: process.id,
                startTime: startExecution,
                duration: executionTime
            });
            
            // Update remaining time
            process.remainingTime -= executionTime;
            currentTime = endExecution;
            
            // If process is not finished, add it back to ready queue
            if (process.remainingTime > 0) {
                // Add any new processes that arrived during execution
                while (processIndex < processData.length && 
                       processData[processIndex].arrival <= currentTime) {
                    readyQueue.push(processData[processIndex]);
                    processIndex++;
                }
                readyQueue.push(process);
            } else {
                // Process is completed
                process.completionTime = currentTime;
                process.turnaroundTime = process.completionTime - process.arrival;
                process.waitingTime = process.turnaroundTime - process.burst;
                
                processResults.push({
                    id: process.id,
                    arrivalTime: process.arrival,
                    burstTime: process.burst,
                    startTime: process.startTime,
                    completionTime: process.completionTime,
                    turnaroundTime: process.turnaroundTime,
                    waitingTime: process.waitingTime,
                    responseTime: process.responseTime
                });
            }
        }
        
        // Calculate metrics
        const totalWaitingTime = processResults.reduce((sum, p) => sum + p.waitingTime, 0);
        const totalTurnaroundTime = processResults.reduce((sum, p) => sum + p.turnaroundTime, 0);
        const totalResponseTime = processResults.reduce((sum, p) => sum + p.responseTime, 0);
        const totalBurstTime = processResults.reduce((sum, p) => sum + p.burstTime, 0);
        
        const avgWaitingTime = totalWaitingTime / processResults.length;
        const avgTurnaroundTime = totalTurnaroundTime / processResults.length;
        const avgResponseTime = totalResponseTime / processResults.length;
        const cpuUtilization = totalBurstTime / currentTime;
        
        return {
            algorithm: 'Round Robin',
            timeQuantum: timeQuantum,
            ganttChart: {
                processes: ganttChart,
                totalTime: currentTime
            },
            processResults: processResults,
            metrics: {
                avgWaitingTime: avgWaitingTime,
                avgTurnaroundTime: avgTurnaroundTime,
                avgResponseTime: avgResponseTime,
                cpuUtilization: cpuUtilization
            }
        };
    }
}

File: VizOs2/VizOs/algorithms/RoundRobin.py
"""
Round Robin Algorithm
Preemptive CPU scheduling algorithm with time quantum
"""

from collections import deque

def execute(processes, quantum):
    """
    Execute Round Robin scheduling algorithm
    
    Args:
        processes: List of process dictionaries
        quantum: Time quantum for each process
        
    Returns:
        Dictionary containing gantt_chart, metrics, and process_details
    """
    # Sort processes by arrival time
    sorted_processes = sorted(processes, key=lambda x: x['arrival_time'])
    
    # Initialize
    gantt_chart = []
    process_details = {p['id']: {
        'id': p['id'],
        'arrival_time': p['arrival_time'],
        'burst_time': p['burst_time'],
        'remaining_time': p['burst_time'],
        'waiting_time': 0,
        'turnaround_time': 0,
        'completion_time': 0
    } for p in sorted_processes}
    
    ready_queue = deque()
    current_time = 0
    process_index = 0
    
    # Main scheduling loop
    while process_index < len(sorted_processes) or ready_queue or any(p['remaining_time'] > 0 for p in process_details.values()):
        # Add arrived processes to ready queue
        while process_index < len(sorted_processes) and sorted_processes[process_index]['arrival_time'] <= current_time:
            proc_id = sorted_processes[process_index]['id']
            ready_queue.append(proc_id)
            process_index += 1
        
        # Get next process from queue
        if ready_queue:
            proc_id = ready_queue.popleft()
            proc = process_details[proc_id]
            
            if proc['remaining_time'] > 0:
                # Execute for quantum or until completion
                execution_time = min(quantum, proc['remaining_time'])
                start_time = current_time
                end_time = current_time + execution_time
                
                # Add to Gantt chart
                gantt_chart.append({
                    'process': proc_id,
                    'start': start_time,
                    'end': end_time
                })
                
                # Update remaining time
                proc['remaining_time'] -= execution_time
                current_time = end_time
                
                # Add arrived processes during execution
                while process_index < len(sorted_processes) and sorted_processes[process_index]['arrival_time'] <= current_time:
                    new_proc_id = sorted_processes[process_index]['id']
                    if new_proc_id not in [p['process'] for p in gantt_chart[-1:]]:
                        ready_queue.append(new_proc_id)
                    process_index += 1
                
                # If process not completed, add back to queue
                if proc['remaining_time'] > 0:
                    ready_queue.append(proc_id)
                else:
                    # Process completed
                    proc['completion_time'] = current_time
                    proc['turnaround_time'] = proc['completion_time'] - proc['arrival_time']
                    proc['waiting_time'] = proc['turnaround_time'] - proc['burst_time']
        else:
            # No process ready, advance time
            if process_index < len(sorted_processes):
                current_time = sorted_processes[process_index]['arrival_time']
            else:
                break
    
    # Calculate metrics
    process_list = list(process_details.values())
    total_waiting_time = sum(p['waiting_time'] for p in process_list)
    total_turnaround_time = sum(p['turnaround_time'] for p in process_list)
    n = len(process_list)
    
    metrics = {
        'average_waiting_time': round(total_waiting_time / n, 2) if n > 0 else 0,
        'average_turnaround_time': round(total_turnaround_time / n, 2) if n > 0 else 0,
        'cpu_utilization': round((sum(p['burst_time'] for p in process_list) / current_time) * 100, 2) if current_time > 0 else 0,
        'throughput': round(n / current_time, 2) if current_time > 0 else 0
    }
    
    return {
        'gantt_chart': gantt_chart,
        'metrics': metrics,
        'process_details': process_list
    }

